apollo-ios/Design/swift-identifiable.md

```diff
@@ -0,0 +1,94 @@
+# Summary
+
+This document proposes a new GraphQL directive named `@apollo_client_ios_identity`, to be used to mark a field which can uniquely identify an object.
+
+# Introduction
+
+SwiftUI makes heavy use of the [Identifiable](https://developer.apple.com/documentation/swift/identifiable) protocol, which is used to track the identity of entities across data changes. If an object with a List is replaced with a different object with the same identity, SwiftUI will animate the item changing instead of animating an insertion. Objects that do not conform to the Identifiable protocol require additional boilerplate to be usable inside SwiftUI.
+
+Apollo Client iOS could assist the developer by adding conformance to the Identifiable protocol to its generated models. Selecting the field to be used as an identity is done by adding a new GraphQL directive.
+
+# Definition
+
+The directive is defined as:
+```graphql
+directive @apollo_client_ios_identity on FIELD | FIELD_DEFINITION
+```
+
+This directive MUST be used on a field with a non-nullable scalar type. 
+
+## Usage in types
+
+A type MAY have a field annotated with the directive.
+
+```graphql
+type Animal {
+	id: ID! @apollo_client_ios_identity
+	name: String
+}
+```
+
+Within a type, the directive MUST NOT be used on more than one field.
+
+## Usage in interfaces
+
+It's possible for interfaces to use the directive on a field.
+
+```graphql
+interface Identifiable {
+	id: ID! @apollo_client_ios_identity
+}
+```
+
+Implementations of this interface MUST copy the directive to the same field.
+
+## Usage in operations
+
+It's likely that an external schema will not use this directive. In this case, an identity MAY be chosen when writing a query. The directive does not need to be included in the query sent to the GraphQL server.
```

Comment:
Are we thinking that applying the directive at the query level would bubble up so that all selections of that type across all queries would receive the `Identifiable`? Wondering if it may result in a confusing result for users if one selection of a given type is Identifiable, while another use of the same type in a different selection set is not?

Reply:
Yea that would be my general sentiment as well, figured this scenario was at least worth discussion because of the behavior being provided. Although since this directive would really leave it up to the user to define how and when to use `Identifiable` then they have the flexibility to apply it to the type itself if they want.

Reply:
> It doesn't seem like a good idea to let queries change the behavior of other queries.

This sentiment might be true if the only application of the directive is `Identifable` conformance; I believe it's larger than that.

Reply:
Can you give an example of how it would be beneficial to look at only the queries, instead of using existing GraphQL constructs like fragments or type extensions? I think it would be surprising behavior if the existence of certain queries affected how other queries are handled.

Reply:
I'm not advocating for operation-based directives to supersede type-based directives - I agree that's bad behaviour.

This RFC should focus on object identity, not `Identifiable` conformance. Once we've expanded it to include all desired applications of the directive there may be a case where its use does influence other selection sets within the same operation. I don't know if those exist, none of us do until we expand the scope of the RFC.